### PR TITLE
Summarize code generation

### PR DESCRIPTION
**Pull request by cilki**
I'm (slowing) trying to summarize each issue in this repository into a cohesive design decision/strategy. I'll start with the one that seems the least controversial.

Request for comment.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,4 +10,21 @@ The redesign of OSHI intends to achieve:
...
+`// TODO`
+
+## Design Decisions
+### [Code Generation](https://github.com/oshi/oshi5/issues/4)
+OSHI's API layer will be automatically generated from a central YAML definitions file. YAML is great for representing tree structures (like OSHI's API) and has a much simpler syntax than JSON or XML. This approach will ensure the resulting API is absolutely uniform, synchronized at all times, and will greatly simplify API changes.
+
+The generator runs on every build (via Maven integration) and updates source files in the API layer's packages according to the central definitions file. Generated files will be committed to the repository.
+
+The definitions file can also be converted into supplemental documentation that contains every feature's description and compatibility requirements in a single place. This addresses the concern brought up [in this comment](https://github.com/oshi/oshi5/issues/2#issuecomment-451220174).
+
```
Comment (dbwiddis) [inline README.md:65]: Looks great as a concept but lacking in details.  Need a couple sentences/bullets outlining what information will be included in the YAML.   In addition to whatever we'll use to represent the name/path/package/json tree/whatever we call it, need to specify the documentation here.

I say this having dealt today with an auto-generated bare-bones API that was utterly useless.

Comment (cilki) [reply README.md:65]: I just added an initial spec for the definitions YAML.
> In addition to whatever we'll use to represent the name/path/package/json tree/whatever we call it, need to specify the documentation here.

I haven't figured out the best way to do this yet.
- Maybe we could have a `package` field applicable to the top-level types in the YAML that describes how users will navigate the API tree to obtain those objects.
- Or just make the top-level types the full paths similar to a Java class. As an example, `Disk` would be `Computer.Hardware.Disk`.

The second option is starting to sound cleaner to me.

File: README.md [hunk 2]
```diff
@@ -21,7 +21,44 @@ OSHI's API layer will be automatically generated from a central YAML definitions
...
+| platforms | A list of the **only** platforms that the attribute is compatible with |
+| incompatible | A list of platforms that the attribute is **not** compatible with |
+| external | A list of external dependencies (like OpenHardwareMonitor) that the attribute requires |
+
+Here's a contrived example displaying all of these elements:
+```
+- Disk
+  - read_bytes: The number of bytes read by the disk
+      type: Long
+      incompatible: sol                          # All platforms except Solaris
```
Comment (dbwiddis) [inline README.md]: I like this. Let's use the full enum names from [JNA's `Platform` class](http://java-native-access.github.io/jna/5.2.0/javadoc/).

File: README.md [hunk 3]
```diff
@@ -21,7 +21,44 @@ OSHI's API layer will be automatically generated from a central YAML definitions
...
+  - partitions: The disk's partitions
+      type: Partition[]                          # Nested container
+  - power_on_time: The disk's total power-on time in hours
+      platforms: lin                             # Only compatible with Linux
+      external: smart                            # Needs S.M.A.R.T. access
```
Comment (dbwiddis) [inline README.md:62]: I like this too. For the *nix commands there is probably a package name involved.  Need a standard code for `root` or `sudo` or `admin` also.

Comment (cilki) [reply README.md:62]: Do you think elevated permissions on Windows and *nix need to be separated like that? I just committed a platform-agnostic field called `permissions` that can be either `elevated` or `none`.

The other option would be to take another list of platforms for `permissions` that defines exactly what platforms require elevated permissions. That would be better than grouping *nix under `sudo`/`root` and Windows under `admin`. I'm not sure if that much control is overkill, though.

Comment (dbwiddis) [reply README.md:62]: Yeah, a boolean on elevated is fine.  On Linux it's possible to change permissions on some directories that are normally root only.  I guess that's "un-elevated".  In any case, too many permutations to track all of them...

**Pull request merged**